Source/WebCore/platform/text/TextFlags.cpp

```diff
@@ -60,48 +58,32 @@ WTF::TextStream& operator<<(TextStream& ts, Kerning kerning)
     return ts;
 }
 
-template <typename T> using BinOp = std::function<T(T, T)>;
-
-template <typename T>
-T reduce(const std::vector<T>& vec, const BinOp<T>& binOp, std::optional<T> initial = { })
-{
-    if (vec.empty())
-        return initial ? *initial : T { };
-
-    auto accumulator = initial ? *initial : vec[0];
-    auto start = initial ? 0 : 1;
-    for (size_t i = start ; i < vec.size() ; i++)
-        accumulator = binOp(accumulator, vec[i]);
-
-    return accumulator;
-}
-
 WTF::TextStream& operator<<(TextStream& ts, FontVariantAlternates alternates)
 {
     if (alternates.isNormal())
         ts << "normal";
     else {
         auto values = alternates.values();
-        std::vector<String> result;
+        Vector<String> result;
         if (values.stylistic)
-            result.push_back("stylistic(" + *values.stylistic + ")");
+            result.append("stylistic(" + *values.stylistic + ")");
         if (values.historicalForms)
-            result.push_back("historical-forms"_s);
+            result.append("historical-forms"_s);
         if (values.styleset)
-            result.push_back("styleset(" + *values.styleset + ")");
+            result.append("styleset(" + *values.styleset + ")");
         if (values.characterVariant)
-            result.push_back("character-variant(" + *values.characterVariant + ")");
+            result.append("character-variant(" + *values.characterVariant + ")");
         if (values.swash)
-            result.push_back("swash(" + *values.swash + ")");
+            result.append("swash(" + *values.swash + ")");
         if (values.ornaments)
-            result.push_back("ornaments(" + *values.ornaments + ")");
+            result.append("ornaments(" + *values.ornaments + ")");
         if (values.annotation)
-            result.push_back("annotation(" + *values.annotation + ")");
+            result.append("annotation(" + *values.annotation + ")");
         
-        BinOp<String> addWithSpace = [](auto a, auto b) { 
+        auto addWithSpace = [](auto a, auto b) { 
             return a + " " + b;
         };
-        ts << reduce(result, addWithSpace);
+        ts << result.reduce(addWithSpace);
```

Comment:
Sadly, I don't think this is a great use case for your reduce implementation. Your implementation is unnecessarily inefficient for Strings and a StringBuilder is recommended for such operations.

Reply:
If you really want to add reduce() to Vector, I recommend finding a better call site for it.

For this call site, we should be using something like this (which we could put in WTFString.h to promote reusability):
```
static String makeStringByJoining(const Vector<String>& strings, UChar separator)
{
    StringBuilder builder;
    for(auto& string : strings) {
        if (!builder.isEmpty())
            builder.append(separator);
        builder.append(string);
    }
    return builder.toString();
}
```

I believe this would be more efficient than your generic reduce implementation.

Or, even more efficient, we don't allocate a Vector at all:
```
StringBuilder builder;
auto appendToBuilder = [&](const String& string) {
    if (!builder.isEmpty())
        builder.append(' ');
    builder.append(string);
};
if (values.stylistic)
     appendToBuilder("stylistic(" + *values.stylistic + ")");
if (values.historicalForms)
    appendToBuilder("historical-forms"_s);
// ...
ts << builder.toString();
```

Reply:
I strongly agree that we must not add code sites that use reduce in a way that results in making tons of temporary strings. And using a vector has the same problem.

For example:
```
appendToBuilder("stylistic(" + *values.stylistic + ")");
```
This is still not a good pattern. It creates and destroys a temporary string; for best efficiency we need each of the pieces to be appended to the builder individually without creating and destroying any temporaries.

For adding separators while using `StringBuilder`, I have been thinking about how to create an object that inserts no separator the first time, but then creates separators later.
```
SeparatorCharacter separator { ' ' };
if (values.stylistic)
    builder.append(separator, "stylistic(", *values.stylistic, ")");
```

Reply:
Yes indeed for this use case `reduce` is less than ideal performance-wise.
I've changed it to use lambda + variadic args, which I think avoid unnecessary string or vector creation/destruction.